File: bin/src/poc/vessel_py/vessel_structured.ts
#!/usr/bin/env bun
/**
 * Vessel for Bun with structured logging - 規約準拠版
 */
import { VesselLogger } from "./vessel_log.ts";

const logger = new VesselLogger("vessel");

async function main() {
  logger.debug("Reading script from stdin");
  const script = await Bun.stdin.text();
  logger.debug(`Script length: ${script.length} chars`);

  try {
    logger.debug("Executing script");
    
    // Create async function for top-level await support
    const AsyncFunction = Object.getPrototypeOf(async function(){}).constructor;
    const func = new AsyncFunction("console", script);
    
    // Wrap console.log to use our logger
    const wrappedConsole = {
      ...console,
      log: (...args: any[]) => logger.output(args.join(" "))
    };
    
    await func(wrappedConsole);
    logger.debug("Script execution completed");
  } catch (error) {
    logger.error("Script execution failed", error);
    process.exit(1);
  }
}

main();

File: bin/src/poc/vessel_py/vessel_structured.py
#!/usr/bin/env python3
"""
Vessel with structured logging - 規約準拠版
"""
import sys
from vessel_log import VesselLogger

# ロガーの初期化
logger = VesselLogger(vessel_type="vessel")

def main():
    # 標準入力からスクリプトを読み込む
    logger.debug("Reading script from stdin")
    script = sys.stdin.read()
    logger.debug(f"Script length: {len(script)} chars")
    
    # 実行環境を準備
    context = {
        '__name__': '__main__',
        'vessel': True,
        # print関数をラップして構造化ログ対応
        'print': logger.output
    }
    
    # スクリプトを実行
    try:
        logger.debug("Executing script")
        exec(script, context)
        logger.debug("Script execution completed")
    except Exception as e:
        logger.error("Script execution failed", error=e)
        sys.exit(1)

if __name__ == "__main__":
    main()